File: frontend/src/types/report.ts
/** Mirrors docs/api-contract.md — Reports section.
 *
 * Generation is asynchronous (Celery + Redis): a report is created in
 * "pending" status, moves to "processing" while the worker aggregates, and
 * only "final" once its totals are computed and stored. "draft" is listed
 * in the contract as unused/predating this flow but kept here so the type
 * stays an exact mirror.
 */
export type ReportStatus = "draft" | "pending" | "processing" | "final";

export interface FacilityBreakdown {
  facility_id: number;
  facility_name: string;
  /** Decimal string, e.g. "12045.30". */
  total_emissions_kg_co2e: string;
}

export interface ReportGenerateRequest {
  organization_id: number;
  report_period_start: string;
  report_period_end: string;
}

/** GET /reports list item — no nested facilities breakdown.
 *
 * `total_emissions_kg_co2e` is `null` until `status` is "final" — a
 * report's totals are computed exactly once by the background worker and
 * stored, never recomputed on read, so there's nothing to show before
 * then.
 */
export interface ReportSummary {
  id: number;
  organization_id: number;
  report_period_start: string;
  report_period_end: string;
  generated_at: string;
  status: ReportStatus;
  total_emissions_kg_co2e: string | null;
}

/** POST /reports/generate and GET /reports/{id} response.
 *
 * `facilities` is `null` under the same condition as
 * `total_emissions_kg_co2e` above (non-"final" status).
 */
export interface Report extends ReportSummary {
  facilities: FacilityBreakdown[] | null;
}

File: backend/app/schemas/report.py
"""Pydantic schemas for the Reports resource.

Matches docs/api-contract.md — Reports section. total_emissions_kg_co2e and
facilities are populated once by the Celery task (app/tasks.py) when a
report reaches FINAL, and stored on the row from then on — not recomputed
live on every read. Both are None while PENDING/PROCESSING.
"""

import enum
from datetime import date, datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel


class ReportStatus(str, enum.Enum):
    """Allowed report statuses (API-level enum, mirrors ReportStatusEnum)."""
    DRAFT = "draft"
    PENDING = "pending"
    PROCESSING = "processing"
    FINAL = "final"


class ReportGenerateRequest(BaseModel):
    """POST /reports/generate request body."""
    organization_id: int
    report_period_start: date
    report_period_end: date


class FacilityBreakdown(BaseModel):
    facility_id: int
    facility_name: str
    total_emissions_kg_co2e: Decimal


class ReportSummaryResponse(BaseModel):
    """GET /reports list item — no nested facilities breakdown."""
    id: int
    organization_id: int
    report_period_start: date
    report_period_end: date
    generated_at: datetime
    status: ReportStatus
    total_emissions_kg_co2e: Optional[Decimal] = None


class ReportDetailResponse(ReportSummaryResponse):
    """POST /reports/generate and GET /reports/{id} response."""
    facilities: Optional[list[FacilityBreakdown]] = None
